File: lib/data/models/npm_package.dart
class NpmPackage {
  final String name;
  final String version;
  final String description;
  final String npmUrl;
  final String homepage;
  final String repositoryUrl;
  final String? readme;
  final DateTime modified;
  final List<String> keywords;
  final int downloads;
  final int stars;
  final String publisher;
  final String? publisherAvatarUrl;
  final String maintainers;
  final List<String> versions;
  final String? latestVersion;
  final String? license;
  final List<String> dependencies;
  final List<String> devDependencies;
  final int openIssues;
  final int openPullRequests;
  final bool hasSecurityVulnerabilities;
  final String? securityPolicyUrl;
  final double trendingScore;
  final String trendingPeriod; // 'daily', 'weekly', 'monthly'
  final DateTime? lastReleaseDate;
  final String? latestReleaseTag;
  final String? latestReleaseName;
  final bool isSubscribed;
  final DateTime subscribedAt;
  final Map<String, dynamic>? metrics;

  NpmPackage({
    required this.name,
    required this.version,
    required this.description,
    required this.npmUrl,
    required this.homepage,
    required this.repositoryUrl,
    this.readme,
    required this.modified,
    required this.keywords,
    required this.downloads,
    this.stars = 0,
    required this.publisher,
    this.publisherAvatarUrl,
    required this.maintainers,
    required this.versions,
    this.latestVersion,
    this.license,
    required this.dependencies,
    required this.devDependencies,
    this.openIssues = 0,
    this.openPullRequests = 0,
    this.hasSecurityVulnerabilities = false,
    this.securityPolicyUrl,
    required this.trendingScore,
    required this.trendingPeriod,
    this.lastReleaseDate,
    this.latestReleaseTag,
    this.latestReleaseName,
    this.isSubscribed = false,
    DateTime? subscribedAt,
    this.metrics,
  }) : subscribedAt = subscribedAt ?? DateTime.now();

  // Factory constructor for API response
  factory NpmPackage.fromJson(Map<String, dynamic> json) {
    return NpmPackage(
      name: json['name'] ?? '',
      version: json['version'] ?? '',
      description: json['description'] ?? '',
      npmUrl: 'https://www.npmjs.com/package/${json['name'] ?? ''}',
      homepage: json['homepage'] ?? '',
      repositoryUrl: json['repository']?['url'] ?? '',
      readme: json['readme'],
      modified: json['modified'] != null
          ? DateTime.tryParse(json['modified']) ?? DateTime.now()
          : DateTime.now(),
      keywords: List<String>.from(json['keywords'] ?? []),
      downloads: json['dist']?['unpackedSize'] ?? 0,
      stars: json['githubStars'] ?? 0,
      publisher: json['publisher']?['username'] ?? '',
      publisherAvatarUrl: json['publisher']?['email'],
      maintainers: _formatMaintainers(json['maintainers'] ?? []),
      versions: json['versions'] != null && json['versions'] is Map
          ? List<String>.from((json['versions'] as Map<String, dynamic>).keys)
          : [],
      latestVersion: json['dist-tags']?['latest'] ?? json['version'],
      license: json['license'],
      dependencies:
          _extractDependencies(json['versions']?['latest']?['dependencies']),
      devDependencies:
          _extractDependencies(json['versions']?['latest']?['devDependencies']),
      openIssues: _extractOpenIssues(json),
      openPullRequests: _extractOpenPullRequests(json),
      hasSecurityVulnerabilities: json['hasSecurityVulnerabilities'] ?? false,
      securityPolicyUrl: json['securityPolicy'],
      trendingScore: _calculateTrendingScore(json),
      trendingPeriod: 'daily',
      lastReleaseDate: _extractLatestReleaseDate(json),
      latestReleaseTag: json['dist-tags']?['latest'] ?? json['version'],
      latestReleaseName: json['name'],
      metrics: json['metrics'],
    );
  }

  // Extract maintainers as formatted string
  static String _formatMaintainers(List<dynamic> maintainers) {
    if (maintainers.isEmpty) return 'No maintainers';
    return maintainers
        .map((m) => m['username'] ?? m['email'] ?? 'Unknown')
        .join(', ');
  }

  // Extract dependencies
  static List<String> _extractDependencies(Map<String, dynamic>? deps) {
    if (deps == null) return [];
    return deps.keys.toList();
  }

  // Extract open issues count
  static int _extractOpenIssues(Map<String, dynamic> json) {
    final repo = json['repository'];
    if (repo is String) {
      // Extract from GitHub URL
      final match = RegExp(r'github\.com/([^/]+/[^/]+)').firstMatch(repo);
      if (match != null) {
        // In a real implementation, you'd make a separate API call to GitHub
        return 0; // Placeholder
      }
    }
    return 0;
  }

  // Extract open pull requests count
  static int _extractOpenPullRequests(Map<String, dynamic> json) {
    // Similar to issues, would need GitHub API integration
    return 0; // Placeholder
  }

  // Extract latest release date
  static DateTime? _extractLatestReleaseDate(Map<String, dynamic> json) {
    final latest = json['versions']?['latest'];
    if (latest != null && latest['time'] != null) {
      return DateTime.parse(latest['time']);
    }
    return null;
  }

  // Calculate trending score
  static double _calculateTrendingScore(Map<String, dynamic> json) {
    final downloads = json['dist']?['unpackedSize'] ?? 0;
    final stars = json['githubStars'] ?? 0;

    final modifiedDate = json['modified'] != null
        ? DateTime.tryParse(json['modified']) ?? DateTime.now()
        : DateTime.now();
    final recentActivity = _calculateRecentActivity(modifiedDate);

    // Weighted scoring algorithm for npm packages
    return (downloads * 0.4) + (stars * 0.4) + (recentActivity * 0.2);
  }

  // Calculate recent activity score
  static double _calculateRecentActivity(DateTime modified) {
    final now = DateTime.now();
    final daysSinceModified = now.difference(modified).inDays;

    if (daysSinceModified <= 1) return 10.0;
    if (daysSinceModified <= 7) return 8.0;
    if (daysSinceModified <= 30) return 6.0;
    if (daysSinceModified <= 90) return 4.0;
    return 2.0;
  }

  // Create copy with updated fields
  NpmPackage copyWith({
    String? name,
    String? version,
    String? description,
    String? npmUrl,
    String? homepage,
    String? repositoryUrl,
    String? readme,
    DateTime? modified,
    List<String>? keywords,
    int? downloads,
    int? stars,
    String? publisher,
    String? publisherAvatarUrl,
    String? maintainers,
    List<String>? versions,
    String? latestVersion,
    String? license,
    List<String>? dependencies,
    List<String>? devDependencies,
    int? openIssues,
    int? openPullRequests,
    bool? hasSecurityVulnerabilities,
    String? securityPolicyUrl,
    double? trendingScore,
    String? trendingPeriod,
    DateTime? lastReleaseDate,
    String? latestReleaseTag,
    String? latestReleaseName,
    bool? isSubscribed,
    DateTime? subscribedAt,
    Map<String, dynamic>? metrics,
  }) {
    return NpmPackage(
      name: name ?? this.name,
      version: version ?? this.version,
      description: description ?? this.description,
      npmUrl: npmUrl ?? this.npmUrl,
      homepage: homepage ?? this.homepage,
      repositoryUrl: repositoryUrl ?? this.repositoryUrl,
      readme: readme ?? this.readme,
      modified: modified ?? this.modified,
      keywords: keywords ?? this.keywords,
      downloads: downloads ?? this.downloads,
      stars: stars ?? this.stars,
      publisher: publisher ?? this.publisher,
      publisherAvatarUrl: publisherAvatarUrl ?? this.publisherAvatarUrl,
      maintainers: maintainers ?? this.maintainers,
      versions: versions ?? this.versions,
      latestVersion: latestVersion ?? this.latestVersion,
      license: license ?? this.license,
      dependencies: dependencies ?? this.dependencies,
      devDependencies: devDependencies ?? this.devDependencies,
      openIssues: openIssues ?? this.openIssues,
      openPullRequests: openPullRequests ?? this.openPullRequests,
      hasSecurityVulnerabilities:
          hasSecurityVulnerabilities ?? this.hasSecurityVulnerabilities,
      securityPolicyUrl: securityPolicyUrl ?? this.securityPolicyUrl,
      trendingScore: trendingScore ?? this.trendingScore,
      trendingPeriod: trendingPeriod ?? this.trendingPeriod,
      lastReleaseDate: lastReleaseDate ?? this.lastReleaseDate,
      latestReleaseTag: latestReleaseTag ?? this.latestReleaseTag,
      latestReleaseName: latestReleaseName ?? this.latestReleaseName,
      isSubscribed: isSubscribed ?? this.isSubscribed,
      subscribedAt: subscribedAt ?? this.subscribedAt,
      metrics: metrics ?? this.metrics,
    );
  }

  // Get formatted downloads string
  String get formattedDownloads {
    if (downloads < 1000) return downloads.toString();
    if (downloads < 1000000) return '${(downloads / 1000).toStringAsFixed(1)}K';
    return '${(downloads / 1000000).toStringAsFixed(1)}M';
  }

  // Get time ago string
  String get timeAgo {
    final now = DateTime.now();
    final difference = now.difference(modified);

    if (difference.inDays > 365) {
      return '${difference.inDays ~/ 365} years ago';
    } else if (difference.inDays > 30) {
      return '${difference.inDays ~/ 30} months ago';
    } else if (difference.inDays > 0) {
      return '${difference.inDays} days ago';
    } else if (difference.inHours > 0) {
      return '${difference.inHours} hours ago';
    } else if (difference.inMinutes > 0) {
      return '${difference.inMinutes} minutes ago';
    } else {
      return 'Just now';
    }
  }

  // Get keywords as formatted string
  String get keywordsString {
    if (keywords.isEmpty) return 'No keywords';
    return keywords.join(', ');
  }

  // Check if has recent activity
  bool get hasRecentActivity {
    final now = DateTime.now();
    final daysSinceModified = now.difference(modified).inDays;
    return daysSinceModified <= 7; // Active in last week
  }

  // Get release info summary
  String get releaseInfo {
    if (latestReleaseTag == null) return 'No releases';
    if (latestReleaseName == null) return 'v$latestReleaseTag';
    return '$latestReleaseTag - $latestReleaseName';
  }

  // Check if package is deprecated
  bool get isDeprecated {
    return version.contains('DEPRECATED') || description.contains('deprecated');
  }

  // Get dependency count
  int get dependencyCount => dependencies.length;

  // Get dev dependency count
  int get devDependencyCount => devDependencies.length;

  // Get total dependency count
  int get totalDependencyCount => dependencyCount + devDependencyCount;

  // Check if has high impact dependencies
  bool get hasHighImpactDependencies {
    final highImpact = ['react', 'angular', 'vue', 'webpack', 'typescript'];
    return dependencies.any((dep) =>
        highImpact.any((impact) => dep.toLowerCase().contains(impact)));
  }

  // Get package size in bytes
  int get packageSize => downloads;

  // Get formatted package size
  String get formattedSize {
    if (downloads < 1024) return '${downloads}B';
    if (downloads < 1024 * 1024) {
      return '${(downloads / 1024).toStringAsFixed(1)}KB';
    }
    return '${(downloads / (1024 * 1024)).toStringAsFixed(1)}MB';
  }

  // JSON serialization
  Map<String, dynamic> toJson() {
    return {
      'name': name,
      'version': version,
      'description': description,
      'npmUrl': npmUrl,
      'homepage': homepage,
      'repositoryUrl': repositoryUrl,
      'readme': readme,
      'modified': modified.toIso8601String(),
      'keywords': keywords,
      'downloads': downloads,
      'stars': stars,
      'publisher': publisher,
      'publisherAvatarUrl': publisherAvatarUrl,
      'maintainers': maintainers,
      'versions': versions,
      'latestVersion': latestVersion,
      'license': license,
      'dependencies': dependencies,
      'devDependencies': devDependencies,
      'openIssues': openIssues,
      'openPullRequests': openPullRequests,
      'hasSecurityVulnerabilities': hasSecurityVulnerabilities,
      'securityPolicyUrl': securityPolicyUrl,
      'trendingScore': trendingScore,
      'trendingPeriod': trendingPeriod,
      'lastReleaseDate': lastReleaseDate?.toIso8601String(),
      'latestReleaseTag': latestReleaseTag,
      'latestReleaseName': latestReleaseName,
      'isSubscribed': isSubscribed,
      'subscribedAt': subscribedAt.toIso8601String(),
      'metrics': metrics,
    };
  }

  @override
  String toString() {
    return 'NpmPackage{name: $name, version: $version, downloads: $downloads}';
  }

  @override
  bool operator ==(Object other) {
    if (identical(this, other)) return true;
    return other is NpmPackage && other.name == name;
  }

  @override
  int get hashCode => name.hashCode;
}

File: packages/app/src/models/NpmPackage.ts
/**
 * npm Package Model
 * Ported from Flutter app: lib/data/models/npm_package.dart
 * Complete model with 30+ fields and helper methods
 */

import {
  getString,
  getStringOrNull,
  getNumber,
  getBoolean,
  getDate,
  getDateOrNull,
  getArray,
  getNestedString,
  getNestedObject,
} from '../utils/typeGuards';

export interface NpmPackageData {
  name: string;
  version: string;
  description: string;
  npmUrl: string;
  homepage: string;
  repositoryUrl: string;
  readme: string | null;
  modified: Date;
  keywords: string[];
  downloads: number;
  stars: number;
  publisher: string;
  publisherAvatarUrl: string | null;
  maintainers: string;
  versions: string[];
  latestVersion: string | null;
  license: string | null;
  dependencies: string[];
  devDependencies: string[];
  openIssues: number;
  openPullRequests: number;
  hasSecurityVulnerabilities: boolean;
  securityPolicyUrl: string | null;
  trendingScore: number;
  trendingPeriod: 'daily' | 'weekly' | 'monthly';
  lastReleaseDate: Date | null;
  latestReleaseTag: string | null;
  latestReleaseName: string | null;
  isSubscribed: boolean;
  subscribedAt: Date;
  metrics: Record<string, unknown> | null;
}

export class NpmPackage implements NpmPackageData {
  name: string;
  version: string;
  description: string;
  npmUrl: string;
  homepage: string;
  repositoryUrl: string;
  readme: string | null;
  modified: Date;
  keywords: string[];
  downloads: number;
  stars: number;
  publisher: string;
  publisherAvatarUrl: string | null;
  maintainers: string;
  versions: string[];
  latestVersion: string | null;
  license: string | null;
  dependencies: string[];
  devDependencies: string[];
  openIssues: number;
  openPullRequests: number;
  hasSecurityVulnerabilities: boolean;
  securityPolicyUrl: string | null;
  trendingScore: number;
  trendingPeriod: 'daily' | 'weekly' | 'monthly';
  lastReleaseDate: Date | null;
  latestReleaseTag: string | null;
  latestReleaseName: string | null;
  isSubscribed: boolean;
  subscribedAt: Date;
  metrics: Record<string, unknown> | null;

  constructor(data: NpmPackageData) {
    Object.assign(this, data);
  }

  /**
   * Create from npm API JSON response
   */
  static fromJSON(json: Record<string, unknown>): NpmPackage {
    const trendingScore = NpmPackage.calculateTrendingScore(json);
    const name = getString(json, 'name');
    const versions = getNestedObject(json, 'versions');
    const versionKeys = versions ? Object.keys(versions) : [];
    const distTags = getNestedObject(json, 'dist-tags');
    const latestTag = distTags ? getString(distTags, 'latest') : null;

    return new NpmPackage({
      name,
      version: getString(json, 'version'),
      description: getString(json, 'description'),
      npmUrl: `https://www.npmjs.com/package/${name}`,
      homepage: getString(json, 'homepage'),
      repositoryUrl: getNestedString(json, 'repository', 'url'),
      readme: getStringOrNull(json, 'readme'),
      modified: getDate(json, 'modified'),
      keywords: getArray<string>(json, 'keywords'),
      downloads: getNumber(getNestedObject(json, 'dist') || {}, 'unpackedSize'),
      stars: getNumber(json, 'githubStars'),
      publisher: getNestedString(json, 'publisher', 'username'),
      publisherAvatarUrl: getNestedString(json, 'publisher', 'email') || null,
      maintainers: NpmPackage.formatMaintainers(getArray(json, 'maintainers')),
      versions: versionKeys,
      latestVersion: latestTag || getString(json, 'version') || null,
      license: getStringOrNull(json, 'license'),
      dependencies: NpmPackage.extractDependencies(
        getNestedObject(getNestedObject(versions || {}, 'latest') || {}, 'dependencies'),
      ),
      devDependencies: NpmPackage.extractDependencies(
        getNestedObject(getNestedObject(versions || {}, 'latest') || {}, 'devDependencies'),
      ),
      openIssues: 0, // Placeholder - would need GitHub API
      openPullRequests: 0, // Placeholder - would need GitHub API
      hasSecurityVulnerabilities: getBoolean(json, 'hasSecurityVulnerabilities'),
      securityPolicyUrl: getStringOrNull(json, 'securityPolicy'),
      trendingScore,
      trendingPeriod: 'daily',
      lastReleaseDate: NpmPackage.extractLatestReleaseDate(json),
      latestReleaseTag: latestTag || getString(json, 'version') || null,
      latestReleaseName: name || null,
      isSubscribed: false,
      subscribedAt: new Date(),
      metrics: getNestedObject(json, 'metrics'),
    });
  }

  /**
   * Format maintainers as string
   */
  private static formatMaintainers(maintainers: unknown[]): string {
    if (!maintainers || maintainers.length === 0) return 'No maintainers';
    return maintainers
      .map(m => {
        if (typeof m === 'object' && m !== null) {
          const obj = m as Record<string, unknown>;
          return getString(obj, 'username') || getString(obj, 'email') || 'Unknown';
        }
        return 'Unknown';
      })
      .join(', ');
  }

  /**
   * Extract dependencies from version object
   */
  private static extractDependencies(deps: Record<string, unknown> | null): string[] {
    if (!deps || typeof deps !== 'object') return [];
    return Object.keys(deps);
  }

  /**
   * Extract latest release date
   */
  private static extractLatestReleaseDate(
    json: Record<string, unknown>,
  ): Date | null {
    const versions = getNestedObject(json, 'versions');
    if (!versions) return null;

    const latest = getNestedObject(versions, 'latest');
    if (!latest) return null;

    return getDateOrNull(latest, 'time');
  }

  /**
   * Calculate trending score (40% downloads, 40% stars, 20% activity)
   */
  private static calculateTrendingScore(json: Record<string, unknown>): number {
    const dist = getNestedObject(json, 'dist');
    const downloads = dist ? getNumber(dist, 'unpackedSize') : 0;
    const stars = getNumber(json, 'githubStars');
    const modifiedDate = getDate(json, 'modified');
    const recentActivity = this.calculateRecentActivity(modifiedDate);

    return downloads * 0.4 + stars * 0.4 + recentActivity * 0.2;
  }

  /**
   * Calculate recent activity score
   */
  private static calculateRecentActivity(modified: Date): number {
    const now = new Date();
    const daysSinceModified = Math.floor(
      (now.getTime() - modified.getTime()) / (1000 * 60 * 60 * 24),
    );

    if (daysSinceModified <= 1) return 10.0;
    if (daysSinceModified <= 7) return 8.0;
    if (daysSinceModified <= 30) return 6.0;
    if (daysSinceModified <= 90) return 4.0;
    return 2.0;
  }

  /**
   * Get formatted downloads string
   */
  get formattedDownloads(): string {
    if (this.downloads < 1000) return this.downloads.toString();
    if (this.downloads < 1000000)
      return `${(this.downloads / 1000).toFixed(1)}K`;
    return `${(this.downloads / 1000000).toFixed(1)}M`;
  }

  /**
   * Get time ago string
   */
  get timeAgo(): string {
    const now = new Date();
    const diffMs = now.getTime() - this.modified.getTime();
    const diffDays = Math.floor(diffMs / (1000 * 60 * 60 * 24));
    const diffHours = Math.floor(diffMs / (1000 * 60 * 60));
    const diffMinutes = Math.floor(diffMs / (1000 * 60));

    if (diffDays > 365) {
      const years = Math.floor(diffDays / 365);
      return `${years} year${years > 1 ? 's' : ''} ago`;
    } else if (diffDays > 30) {
      const months = Math.floor(diffDays / 30);
      return `${months} month${months > 1 ? 's' : ''} ago`;
    } else if (diffDays > 0) {
      return `${diffDays} day${diffDays > 1 ? 's' : ''} ago`;
    } else if (diffHours > 0) {
      return `${diffHours} hour${diffHours > 1 ? 's' : ''} ago`;
    } else if (diffMinutes > 0) {
      return `${diffMinutes} minute${diffMinutes > 1 ? 's' : ''} ago`;
    } else {
      return 'Just now';
    }
  }

  /**
   * Get keywords as formatted string
   */
  get keywordsString(): string {
    if (this.keywords.length === 0) return 'No keywords';
    return this.keywords.join(', ');
  }

  /**
   * Check if has recent activity (last 7 days)
   */
  get hasRecentActivity(): boolean {
    const now = new Date();
    const daysSinceModified = Math.floor(
      (now.getTime() - this.modified.getTime()) / (1000 * 60 * 60 * 24),
    );
    return daysSinceModified <= 7;
  }

  /**
   * Get release info summary
   */
  get releaseInfo(): string {
    if (!this.latestReleaseTag) return 'No releases';
    if (!this.latestReleaseName) return `v${this.latestReleaseTag}`;
    return `${this.latestReleaseTag} - ${this.latestReleaseName}`;
  }

  /**
   * Check if package is deprecated
   */
  get isDeprecated(): boolean {
    return (
      this.version.includes('DEPRECATED') ||
      this.description.toLowerCase().includes('deprecated')
    );
  }

  /**
   * Get dependency count
   */
  get dependencyCount(): number {
    return this.dependencies.length;
  }

  /**
   * Get dev dependency count
   */
  get devDependencyCount(): number {
    return this.devDependencies.length;
  }

  /**
   * Get total dependency count
   */
  get totalDependencyCount(): number {
    return this.dependencyCount + this.devDependencyCount;
  }

  /**
   * Check if has high impact dependencies
   */
  get hasHighImpactDependencies(): boolean {
    const highImpact = ['react', 'angular', 'vue', 'webpack', 'typescript'];
    return this.dependencies.some(dep =>
      highImpact.some(impact => dep.toLowerCase().includes(impact)),
    );
  }

  /**
   * Get formatted package size
   */
  get formattedSize(): string {
    if (this.downloads < 1024) return `${this.downloads}B`;
    if (this.downloads < 1024 * 1024)
      return `${(this.downloads / 1024).toFixed(1)}KB`;
    return `${(this.downloads / (1024 * 1024)).toFixed(1)}MB`;
  }

  /**
   * Convert to JSON
   */
  toJSON(): Record<string, unknown> {
    return {
      name: this.name,
      version: this.version,
      description: this.description,
      npmUrl: this.npmUrl,
      homepage: this.homepage,
      repositoryUrl: this.repositoryUrl,
      readme: this.readme,
      modified: this.modified.toISOString(),
      keywords: this.keywords,
      downloads: this.downloads,
      stars: this.stars,
      publisher: this.publisher,
      publisherAvatarUrl: this.publisherAvatarUrl,
      maintainers: this.maintainers,
      versions: this.versions,
      latestVersion: this.latestVersion,
      license: this.license,
      dependencies: this.dependencies,
      devDependencies: this.devDependencies,
      openIssues: this.openIssues,
      openPullRequests: this.openPullRequests,
      hasSecurityVulnerabilities: this.hasSecurityVulnerabilities,
      securityPolicyUrl: this.securityPolicyUrl,
      trendingScore: this.trendingScore,
      trendingPeriod: this.trendingPeriod,
      lastReleaseDate: this.lastReleaseDate?.toISOString(),
      latestReleaseTag: this.latestReleaseTag,
      latestReleaseName: this.latestReleaseName,
      isSubscribed: this.isSubscribed,
      subscribedAt: this.subscribedAt.toISOString(),
      metrics: this.metrics,
    };
  }

  /**
   * Create copy with updated fields
   */
  copyWith(updates: Partial<NpmPackageData>): NpmPackage {
    return new NpmPackage({ ...this, ...updates });
  }
}

export default NpmPackage;
